packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,191 @@
+            ? Infinity : this.defaultAttachFlowSoftLimitInBytes;
+
+        this.attachFlowBatch = new BatchManager({
+            hardLimit,
+            softLimit,
+        });
+        this.mainBatch = new BatchManager({
+            hardLimit
+        });
+    }
+
+    public get isEmpty(): boolean {
+        return this.attachFlowBatch.length === 0 && this.mainBatch.length === 0;
+    }
+
+    public submit(message: BatchMessage) {
+        if (!this.mainBatch.push(message)) {
+            throw new GenericError(
+                "BatchTooLarge",
+                /* error */ undefined,
+                {
+                    opSize: (message.contents?.length) ?? 0,
+                    count: this.mainBatch.length,
+                    limit: this.mainBatch.options.hardLimit,
+                });
+        }
+    }
+
+    public submitAttach(message: BatchMessage) {
+        if (!this.attachFlowBatch.push(message)) {
+            // BatchManager has two limits - soft limit & hard limit. Soft limit is only engaged
+            // when queue is not empty.
+            // Flush queue & retry. Failure on retry would mean - single message is bigger than hard limit
+            this.flushInternal(this.attachFlowBatch.popBatch());
+            if (!this.attachFlowBatch.push(message)) {
+                throw new GenericError(
+                    "BatchTooLarge",
+                    /* error */ undefined,
+                    {
+                        opSize: (message.contents?.length) ?? 0,
+                        count: this.attachFlowBatch.length,
+                        limit: this.attachFlowBatch.options.hardLimit,
+                    });
+            }
+        }
+    }
+
+    public flush() {
+        this.flushInternal(this.attachFlowBatch.popBatch());
+        this.flushInternal(this.mainBatch.popBatch());
+    }
+
+    private flushInternal(batch: IBatch) {
+        this.flushBatch(
+            this.addBatchMetadataToBatch(
+                this.processBatch(batch)));
+    }
+
+    private addBatchMetadataToBatch(batch: IBatch): IBatch {
+        if (batch.content.length > 1) {
+            batch.content[0].metadata = {
+                ...batch.content[0].metadata,
+                batch: true
+            };
+            batch.content[batch.content.length - 1].metadata = {
+                ...batch.content[batch.content.length - 1].metadata,
+                batch: false
+            };
+        }
+
+        return batch;
+    }
+
+    private processBatch(batch: IBatch): IBatch {
+        if (batch.content.length === 0 || this.options.compressionOptions === undefined) {
+            return batch;
+        }
+
+        let processedBatch = batch;
+        if (this.options.compressionOptions.minimumBatchSizeInBytes < batch.contentSizeInBytes) {
+            processedBatch = this.batchProcessors.compressor.processOutgoing(batch);
+        }
+
+        return processedBatch;
+    }
+
+    private flushBatch(batch: IBatch): void {
```

Comment:
I'd call it compressBatch().

Reply:
For now it probably makes sense as it's the only thing it might do. But this is where chunking will also be called. I'll rename these to something more relevant.